### PR TITLE
Thermostat Fixes

### PR DESCRIPTION
**Pull request by jwhitak**
# They were broken?
yes. yes they were.

## What this does
Fixes thermostats so that their controls actually do something now besides make an indefinite clicking noise. The following are a list of the changes:
* Thermostat controls actually set the temperature now, properly setting and using the correct variable.
image
(@brndd's one mistake, uncaught for 5 years)

* Correctly shows the maximum amounts in C instead of K.
image

* Rounds off odd decimal inputs.
Before:
image
After:
image

* Allows inputing the exact minimum and maximum, such as the default 0 and 40C.
Before:
image
After:
This doesn't happen.

* Allows inputting 0C as a valid input.

* BONUS: If you try to input a number outside of the range, it'll be clamped to the range. E.g. if you try to drop 99999 in a default thermostat, it'll treat it as 40.

* Thermostats will now announce if they switch between heating and cooling modes.
image

## Why it's good
I want it warm in my leather making hotroom (thermostats set to 40C, the default maximum, can't get the room warm enough to cure leather quickly...), and I want it cold in the kitchen backroom, like it should be.

## How it was tested
The usual - opening up ye olde test server and playing around with all the changes. Heating and cooling arbitarily, trying to set it to 0C to freeze to death, then changing the threshold presets to let me really up the thermostat and cure leather.
Before:
image
After:
image

## Changelog

:cl:
 * bugfix: Thermostats now properly work when setting new temperatures.

### CODE REVIEW

**Comment by SECBATON-GRIFFON** [comment]
who touched the thermostat

**Comment by toomykins** [comment]
amazingly good fix.
hope this also miraculously fixes #35711

File: code/game/machinery/alarm.dm [hunk 1]
```diff
@@ -976,13 +989,13 @@ var/global/list/air_alarms = list()
 		else
 			max_temperature = temperature_threshold.max_1() - T0C
 			min_temperature = temperature_threshold.min_1() - T0C
-		var/input_temperature = input("What temperature (in C) would you like the system to target? (Capped between [min_temperature]C and [max_temperature]C).\n\nNote that the cooling unit in this air alarm can not go below [MIN_TEMPERATURE]C or above [MAX_TEMPERATURE]C by itself. ", "Thermostat Controls") as num|null
+		var/input_temperature = input("What temperature (in C) would you like the system to target? (Capped between [min_temperature]C and [max_temperature]C).\n\nNote that the cooling unit in this air alarm can not go below [MIN_TEMPERATURE - T0C]C or above [MAX_TEMPERATURE - T0C]C by itself. ", "Thermostat Controls") as num|null
 		if(input_temperature==null)
 			return 1
-		if(!input_temperature || input_temperature >= max_temperature || input_temperature <= min_temperature)
+		if(input_temperature > max_temperature || input_temperature < min_temperature)
```
Comment (west3436) [inline code/game/machinery/alarm.dm]: Could also just allow values outside the range but clamp the result. Atmos computers work like this.

File: code/game/machinery/alarm.dm [hunk 2]
```diff
@@ -976,13 +989,13 @@ var/global/list/air_alarms = list()
 		else
 			max_temperature = temperature_threshold.max_1() - T0C
 			min_temperature = temperature_threshold.min_1() - T0C
-		var/input_temperature = input("What temperature (in C) would you like the system to target? (Capped between [min_temperature]C and [max_temperature]C).\n\nNote that the cooling unit in this air alarm can not go below [MIN_TEMPERATURE]C or above [MAX_TEMPERATURE]C by itself. ", "Thermostat Controls") as num|null
+		var/input_temperature = input("What temperature (in C) would you like the system to target? (Capped between [min_temperature]C and [max_temperature]C).\n\nNote that the cooling unit in this air alarm can not go below [MIN_TEMPERATURE - T0C]C or above [MAX_TEMPERATURE - T0C]C by itself. ", "Thermostat Controls") as num|null
```
Comment (west3436) [inline code/game/machinery/alarm.dm:992]:
```suggestion
		var/input_temperature = input("What temperature (in C) would you like the system to target? (Capped between [min_temperature]C and [max_temperature]C).\n\nNote that the cooling unit in this air alarm can not go below [MIN_TEMPERATURE - T0C]C or above [MAX_TEMPERATURE + T0C]C by itself. ", "Thermostat Controls") as num|null
```

File: code/game/machinery/computer/atmos_control.dm [hunk 3]
```diff
@@ -427,13 +427,13 @@ var/global/list/atmos_controllers = list()
 		if(href_list["set_preset_setting"] == "target_temperature")
 			var/max_temperature = MAX_TARGET_TEMPERATURE - T0C //these defines should come from code\game\machinery\alarm.dm
 			var/min_temperature = MIN_TARGET_TEMPERATURE - T0C
-			var/input_temperature = input("What temperature (in C) would you like the system to target? (Capped between [min_temperature]C and [max_temperature]C).\n\nNote that the cooling unit in this air alarm can not go below [MIN_TEMPERATURE]C or above [MAX_TEMPERATURE]C by itself. ", "Thermostat Controls") as num|null
+			var/input_temperature = input("What temperature (in C) would you like the system to target? (Capped between [min_temperature]C and [max_temperature]C).\n\nNote that the cooling unit in this air alarm can not go below [MIN_TEMPERATURE - T0C]C or above [MAX_TEMPERATURE - T0C]C by itself. ", "Thermostat Controls") as num|null
```
Comment (west3436) [inline code/game/machinery/computer/atmos_control.dm:430]:
```suggestion
			var/input_temperature = input("What temperature (in C) would you like the system to target? (Capped between [min_temperature]C and [max_temperature]C).\n\nNote that the cooling unit in this air alarm can not go below [MIN_TEMPERATURE - T0C]C or above [MAX_TEMPERATURE + T0C]C by itself. ", "Thermostat Controls") as num|null
```

File: code/game/machinery/computer/atmos_control.dm [hunk 4]
```diff
@@ -613,13 +613,13 @@ var/global/list/atmos_controllers = list()
 			else
 				max_temperature = temperature_threshold.max_1() - T0C
 				min_temperature = temperature_threshold.min_1() - T0C
-			var/input_temperature = input("What temperature (in C) would you like the system to target? (Capped between [min_temperature]C and [max_temperature]C).\n\nNote that the cooling unit in this air alarm can not go below [MIN_TEMPERATURE]C or above [MAX_TEMPERATURE]C by itself. ", "Thermostat Controls") as num|null
+			var/input_temperature = input("What temperature (in C) would you like the system to target? (Capped between [min_temperature]C and [max_temperature]C).\n\nNote that the cooling unit in this air alarm can not go below [MIN_TEMPERATURE - T0C]C or above [MAX_TEMPERATURE - T0C]C by itself. ", "Thermostat Controls") as num|null
```
Comment (west3436) [inline code/game/machinery/computer/atmos_control.dm:613]:
```suggestion
			var/input_temperature = input("What temperature (in C) would you like the system to target? (Capped between [min_temperature]C and [max_temperature]C).\n\nNote that the cooling unit in this air alarm can not go below [MIN_TEMPERATURE - T0C]C or above [MAX_TEMPERATURE + T0C]C by itself. ", "Thermostat Controls") as num|null
```

**Review by west3436 — CHANGES_REQUESTED** [review]
What is the default control range (max-min) for these? Should be wide enough to prevent constant triggering.

File: code/game/machinery/alarm.dm (hunk 2 above)
Comment (jwhitak) [reply code/game/machinery/alarm.dm:992]:
```
#define MAX_TEMPERATURE T0C+90
#define MIN_TEMPERATURE T0C-40
```
Which is 90C and -40C.
So, to confirm that the math is right...
(90 + T0C) = 90C (in K)
(-40 + T0C) = -40C (in K)
Adding T0C to a C number converts it to K. So, to convert from K to C, you must subtract T0C as follows:
(-40 + T0C) - T0C = -40C
(90 + T0C) - T0C = 90C
vs the proposed change,
(90 + T0C) + T0C = considerably higher than 90C

Therefore the original line is correct.

Bonus: note the image of the correctly displayed numbers in the PR body

File: code/game/machinery/alarm.dm (hunk 1 above)
Comment (jwhitak) [reply code/game/machinery/alarm.dm]: Agreed.

File: code/game/machinery/computer/atmos_control.dm (hunk 3 above)
Comment (jwhitak) [reply code/game/machinery/computer/atmos_control.dm:430]: See other comment

File: code/game/machinery/computer/atmos_control.dm (hunk 4 above)
Comment (jwhitak) [reply code/game/machinery/computer/atmos_control.dm:613]: See other comment

File: code/game/machinery/alarm.dm (hunk 1 above)
Comment (jwhitak) [reply code/game/machinery/alarm.dm]: Implemented, tested, and committed! (both negative and positive direction work)
image

**Comment by jwhitak** [comment]
The control range is that the air alarm's heating/cooling will turn on if it exceeds 2C of the desired temperature, and turn off if it's within 0.5C of the desired temperature. The amount heated/cooled scales based on the difference in desired/actual temperature, and the heat capacity of the gas sample taken.

**Review by west3436 — APPROVED** [review]

**Pull request merged**